File: test/python/disbursement/transfer_to_dana_inquiry_status_test.py
import os
import pytest
import time
from dana.utils.snap_configuration import SnapConfiguration, AuthSettings, Env
from dana.disbursement.v1.enum import *
from dana.disbursement.v1.models import *
from dana.disbursement.v1 import *
from dana.disbursement.v1.api import *
from dana.api_client import ApiClient
from dana.exceptions import *
from uuid import uuid4
from helper.util import get_request, with_delay
from helper.api_helpers import get_headers_with_signature, execute_and_assert_api_error
from helper.assertion import assert_response, assert_fail_response

transfer_to_dana_title_case = "TransferToDana"
title_case = "TransferToDanaInquiryStatus"
json_path_file = "resource/request/components/Disbursement.json"

configuration = SnapConfiguration(
    api_key=AuthSettings(
        PRIVATE_KEY=os.environ.get("PRIVATE_KEY"),
        ORIGIN=os.environ.get("ORIGIN"),
        X_PARTNER_ID=os.environ.get("X_PARTNER_ID"),
        DANA_ENV=Env.SANDBOX,
        X_DEBUG="true",
        CLIENT_SECRET=os.environ.get("CLIENT_SECRET"),
    )
)

with ApiClient(configuration) as api_client:
    api_instance = DisbursementApi(api_client)

# Shared test data for cross-test dependencies
original_partner_reference_paid = None
original_partner_reference_failed = None

def create_disbursement_paid():
    """Creates a successful disbursement transaction for testing status inquiry"""
    global original_partner_reference_paid
    
    disbursement_request = get_request(json_path_file, "TransferToDana", "TopUpCustomerValid")
    original_partner_reference_paid = str(uuid4())
    disbursement_request["partnerReferenceNo"] = original_partner_reference_paid
    
    request_obj = TransferToDanaRequest.from_dict(disbursement_request)
    api_instance.transfer_to_dana(request_obj)

def create_disbursement_failed():
    """Creates a failed disbursement transaction for testing status inquiry"""
    global original_partner_reference_failed
    
    disbursement_request = get_request(json_path_file, "TransferToDana", "TopUpCustomerExceedAmountLimit")
    original_partner_reference_failed = str(uuid4())
    disbursement_request["partnerReferenceNo"] = original_partner_reference_failed
    
    request_obj = TransferToDanaRequest.from_dict(disbursement_request)
    api_instance.transfer_to_dana(request_obj)

def setup_module():
    """Test setup hook that creates prerequisite disbursement transactions"""
    try:
        create_disbursement_paid()
        print(f"Shared order created with reference: {original_partner_reference_paid}")
        create_disbursement_failed()
        print(f"Shared order created with reference: {original_partner_reference_failed}")
    except Exception as e:
        print(f"Failed to create shared order - tests cannot continue: {e}")

@with_delay()
def test_inquiry_topup_status_valid_paid():
    case_name = "InquiryTopUpStatusValidPaid"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["originalPartnerReferenceNo"] = original_partner_reference_paid
    request_obj = TransferToDanaInquiryStatusRequest.from_dict(json_dict)

    try:
        response = api_instance.transfer_to_dana_inquiry_status(request_obj)
        
        variable_dict = {
            'originalPartnerReferenceNo': original_partner_reference_paid,
            'originalReferenceNo': response.original_reference_no
        }
        
        assert_response(json_path_file, title_case, case_name, TransferToDanaInquiryStatusResponse.to_json(response), variable_dict)

    except ApiException as e:
        pytest.fail(f"API call failed: {e}")

@with_delay()
def test_inquiry_topup_status_valid_failed():
    case_name = "InquiryTopUpStatusValidFail"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["originalPartnerReferenceNo"] = original_partner_reference_failed
    request_obj = TransferToDanaInquiryStatusRequest.from_dict(json_dict)

    try:
        response = api_instance.transfer_to_dana_inquiry_status(request_obj)
        
        variable_dict = {
            'originalPartnerReferenceNo': original_partner_reference_failed,
            'originalReferenceNo': response.original_reference_no
        }
        
        assert_response(json_path_file, title_case, case_name, TransferToDanaInquiryStatusResponse.to_json(response), variable_dict)

    except ApiException as e:
        pytest.fail(f"API call failed: {e}")

@with_delay()
def test_inquiry_topup_status_invalid_field_format():
    case_name = "InquiryTopUpStatusInvalidFieldFormat"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["originalPartnerReferenceNo"] = original_partner_reference_paid
    request_obj = TransferToDanaInquiryStatusRequest.from_dict(json_dict)

    try:
        api_instance.transfer_to_dana_inquiry_status(request_obj)
        pytest.fail("Expected ApiException for invalid field format but the API call succeeded")

    except ApiException as e:
        assert_fail_response(json_path_file, title_case, case_name, str(e.body), 
                           {'partnerReferenceNo': original_partner_reference_paid})

@with_delay()
def test_inquiry_topup_status_not_found_transaction():
    case_name = "InquiryTopUpStatusNotFoundTransaction"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["originalPartnerReferenceNo"] = "test123123"
    request_obj = TransferToDanaInquiryStatusRequest.from_dict(json_dict)

    try:
        api_instance.transfer_to_dana_inquiry_status(request_obj)
        pytest.fail("Expected ApiException for transaction not found but the API call succeeded")

    except ApiException as e:
        assert_fail_response(json_path_file, title_case, case_name, str(e.body), 
                           {'originalPartnerReferenceNo': "test123123"})

@with_delay()
def test_inquiry_topup_status_missing_mandatory_field():
    case_name = "InquiryTopUpStatusMissingMandatoryField"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["originalPartnerReferenceNo"] = original_partner_reference_paid
    request_obj = TransferToDanaInquiryStatusRequest.from_dict(json_dict)

    headers = get_headers_with_signature(
        method="POST",
        resource_path="v1.0/emoney/topup-status.htm",
        request_obj=json_dict,
        with_timestamp=True,
        invalid_timestamp=False
    )
    
    headers["X-TIMESTAMP"] = ""

    execute_and_assert_api_error(
        api_client,
        "POST",
        "https://api.sandbox.dana.id/rest/v1.0/emoney/topup-status",
        request_obj,
        headers,
        400,
        json_path_file,
        title_case,
        case_name
    )

@with_delay()
def test_inquiry_topup_status_unauthorized_signature():
    case_name = "InquiryTopUpStatusUnauthorizedSignature"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["originalPartnerReferenceNo"] = original_partner_reference_paid
    request_obj = TransferToDanaInquiryStatusRequest.from_dict(json_dict)

    headers = get_headers_with_signature(
        method="POST",
        resource_path="v1.0/emoney/topup-status.htm",
        request_obj=json_dict,
        with_timestamp=True,
        invalid_timestamp=False
    )
    
    headers["X-SIGNATURE"] = "85be817c55b2c135157c7e89f52499bf0c25ad6eeebe04a986e8c862561b19a5"

    execute_and_assert_api_error(
        api_client,
        "POST",
        "https://api.sandbox.dana.id/rest/v1.0/emoney/topup-status",
        request_obj,
        headers,
        401,
        json_path_file,
        title_case,
        case_name
    )

File: test/php/disbursement/TransferToDanaInquiryStatusTest.php
<?php

namespace DanaUat\Disbursement;

use PHPUnit\Framework\TestCase;
use Dana\Disbursement\v1\Api\v1\Api\DisbursementApi;
use Dana\ObjectSerializer;
use Dana\Configuration;
use Dana\Env;
use Dana\ApiException;
use Dana\Disbursement\v1\Api\DisbursementApi as ApiDisbursementApi;
use DanaUat\Helper\Assertion;
use DanaUat\Helper\Util;
use Exception;

class TransferToDanaInquiryStatusTest extends TestCase
{
    private static $titleCase = 'TransferToDanaInquiryStatus';
    private static $jsonPathFile = 'resource/request/components/Disbursement.json';
    private static $apiInstance;
    private static $merchantId;
    private static $originalPartnerReferencePaid;
    private static $originalPartnerReferenceFailed;

    public static function setUpBeforeClass(): void
    {
        // Set up configuration with authentication settings
        $configuration = new Configuration();

        // The Configuration constructor automatically loads values from environment variables
        // But we can manually set them if needed
        $configuration->setApiKey('PRIVATE_KEY', getenv('PRIVATE_KEY'));
        $configuration->setApiKey('ORIGIN', getenv('ORIGIN'));
        $configuration->setApiKey('X_PARTNER_ID', getenv('X_PARTNER_ID'));
        $configuration->setApiKey('ENV', Env::SANDBOX);

        // Create DisbursementApi instance directly with configuration
        self::$apiInstance = new ApiDisbursementApi(null, $configuration);

        // Store merchantId for reuse
        self::$merchantId = getenv('MERCHANT_ID');

        // Create prerequisite disbursement transactions for status inquiry tests
        self::createDisbursementPaid();
        self::createDisbursementFailed();
    }

    /**
     * Helper function to create a successful disbursement for status testing
     */
    private static function createDisbursementPaid(): void
    {
        try {
            // Get the request data for a valid transfer
            $jsonDict = Util::getRequest(
                'resource/request/components/Disbursement.json',
                'TransferToDana',
                'TopUpCustomerValid'
            );

            // Generate unique partner reference number
            self::$originalPartnerReferencePaid = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = self::$originalPartnerReferencePaid;

            // Create request object
            $transferToDanaRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\TransferToDanaRequest'
            );

            // Make the API call to create a successful transaction
            self::$apiInstance->transferToDana($transferToDanaRequestObj);
            
            echo "Shared paid order created with reference: " . self::$originalPartnerReferencePaid . "\n";
        } catch (Exception $e) {
            echo "Failed to create shared paid order: " . $e->getMessage() . "\n";
        }
    }

    /**
     * Helper function to create a failed disbursement for status testing
     */
    private static function createDisbursementFailed(): void
    {
        try {
            // Get the request data for an amount limit exceeded transfer
            $jsonDict = Util::getRequest(
                'resource/request/components/Disbursement.json',
                'TransferToDana',
                'TopUpCustomerExceedAmountLimit'
            );

            // Generate unique partner reference number
            self::$originalPartnerReferenceFailed = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = self::$originalPartnerReferenceFailed;

            // Create request object
            $transferToDanaRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\TransferToDanaRequest'
            );

            // Make the API call to create a failed transaction
            try {
                self::$apiInstance->transferToDana($transferToDanaRequestObj);
            } catch (ApiException $e) {
                // Expected to fail - this creates a failed transaction
                echo "Shared failed order created with reference: " . self::$originalPartnerReferenceFailed . "\n";
            }
        } catch (Exception $e) {
            echo "Failed to create shared failed order: " . $e->getMessage() . "\n";
        }
    }

    /**
     * Should successfully inquire transfer to DANA status PAID
     */
    public function testInquiryTopUpStatusValidPaid(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryTopUpStatusValidPaid';

            try {
                // Get and prepare the request
                $jsonDict = Util::getRequest(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName
                );

                // Use the original partner reference from the successful transaction
                $jsonDict['originalPartnerReferenceNo'] = self::$originalPartnerReferencePaid;

                // Create a TransferToDanaInquiryStatusRequest object from the JSON request data
                $inquiryStatusRequestObj = ObjectSerializer::deserialize(
                    $jsonDict,
                    'Dana\Disbursement\v1\Model\TransferToDanaInquiryStatusRequest'
                );

                // Make the API call
                $apiResponse = self::$apiInstance->transferToDanaInquiryStatus($inquiryStatusRequestObj);

                // Assert the response matches the expected data
                Assertion::assertResponse(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName,
                    $apiResponse->__toString(),
                    [
                        'originalPartnerReferenceNo' => self::$originalPartnerReferencePaid,
                        'originalReferenceNo' => $apiResponse->getOriginalReferenceNo()
                    ]
                );

                $this->assertTrue(true);
            } catch (ApiException $e) {
                $this->fail('Failed to inquire transfer to DANA status: ' . $e->getMessage());
            } catch (Exception $e) {
                $this->fail('Unexpected exception: ' . $e->getMessage());
            }
        });
    }

    /**
     * Should successfully inquire transfer to DANA status FAILED
     */
    public function testInquiryTopUpStatusValidFailed(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryTopUpStatusValidFail';

            try {
                // Get and prepare the request
                $jsonDict = Util::getRequest(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName
                );

                // Use the original partner reference from the failed transaction
                $jsonDict['originalPartnerReferenceNo'] = self::$originalPartnerReferenceFailed;

                // Create a TransferToDanaInquiryStatusRequest object from the JSON request data
                $inquiryStatusRequestObj = ObjectSerializer::deserialize(
                    $jsonDict,
                    'Dana\Disbursement\v1\Model\TransferToDanaInquiryStatusRequest'
                );

                // Make the API call
                $apiResponse = self::$apiInstance->transferToDanaInquiryStatus($inquiryStatusRequestObj);

                // Assert the response matches the expected data
                Assertion::assertResponse(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName,
                    $apiResponse->__toString(),
                    [
                        'originalPartnerReferenceNo' => self::$originalPartnerReferenceFailed,
                        'originalReferenceNo' => $apiResponse->getOriginalReferenceNo()
                    ]
                );

                $this->assertTrue(true);
            } catch (ApiException $e) {
                $this->fail('Failed to inquire transfer to DANA status: ' . $e->getMessage());
            } catch (Exception $e) {
                $this->fail('Unexpected exception: ' . $e->getMessage());
            }
        });
    }

    /**
     * Should fail inquiry when input is invalid
     */
    public function testInquiryTopUpStatusInvalidFieldFormat(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryTopUpStatusInvalidFieldFormat';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Use the original partner reference from the successful transaction
            $jsonDict['originalPartnerReferenceNo'] = self::$originalPartnerReferencePaid;

            // Create a TransferToDanaInquiryStatusRequest object from the JSON request data
            $inquiryStatusRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\TransferToDanaInquiryStatusRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->transferToDanaInquiryStatus($inquiryStatusRequestObj);

                $this->fail('Expected ApiException for invalid field format but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 400 Bad Request for invalid field format
                $this->assertEquals(400, $e->getCode(), "Expected HTTP 400 Bad Request for invalid field format, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['partnerReferenceNo' => self::$originalPartnerReferencePaid]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }

    /**
     * Should fail inquiry when transaction not found
     */
    public function testInquiryTopUpStatusNotFoundTransaction(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryTopUpStatusNotFoundTransaction';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Use a non-existent partner reference
            $nonExistentRef = 'test123123';
            $jsonDict['originalPartnerReferenceNo'] = $nonExistentRef;

            // Create a TransferToDanaInquiryStatusRequest object from the JSON request data
            $inquiryStatusRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\TransferToDanaInquiryStatusRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->transferToDanaInquiryStatus($inquiryStatusRequestObj);

                $this->fail('Expected ApiException for transaction not found but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 404 Not Found for transaction not found
                $this->assertEquals(404, $e->getCode(), "Expected HTTP 404 Not Found for transaction not found, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['originalPartnerReferenceNo' => $nonExistentRef]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }

    /**
     * Should fail inquiry when transaction is missing mandatory field
     */
    public function testInquiryTopUpStatusMissingMandatoryField(): void
    {
        try {
            Util::withDelay(function() {
                $caseName = 'InquiryTopUpStatusMissingMandatoryField';
                
                // Get the request data from the JSON file
                $requestData = Util::getRequest(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName
                );
                
                // Use the original partner reference from the successful transaction
                $requestData['originalPartnerReferenceNo'] = self::$originalPartnerReferencePaid;
                
                // Create headers without timestamp to test validation
                $headers = Util::getHeadersWithSignature(
                    'POST', 
                    '/rest/v1.0/emoney/topup-status',
                    $requestData,
                    false  // withTimestamp = false to test missing X-TIMESTAMP
                );
                
                // Make direct API call with missing timestamp
                try {
                    Util::executeApiRequest(
                        'POST',
                        'https://api.sandbox.dana.id/rest/v1.0/emoney/topup-status',
                        $headers,
                        $requestData
                    );
                    
                    $this->fail('Expected ApiException for missing mandatory field but the API call succeeded');
                } catch (ApiException $e) {
                    // We expect a 400 Bad Request for missing mandatory field
                    $this->assertEquals(400, $e->getCode(), "Expected HTTP 400 Bad Request for missing mandatory field, got {$e->getCode()}");

                    // Get the response body from the exception
                    $responseContent = (string)$e->getResponseBody();
                    
                    // Use assertFailResponse to validate the error response
                    Assertion::assertFailResponse(
                        self::$jsonPathFile, 
                        self::$titleCase, 
                        $caseName, 
                        $responseContent,
                        ['partnerReferenceNo' => self::$originalPartnerReferencePaid]
                    );
                } catch (Exception $e) {
                    $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
                }
            });
        } catch (Exception $e) {            
            $this->fail("Unexpected exception: " . $e->getMessage());
        }
    }

    /**
     * Should fail inquiry when signature is unauthorized
     */
    public function testInquiryTopUpStatusUnauthorizedSignature(): void
    {
        try {
            Util::withDelay(function() {
                $caseName = 'InquiryTopUpStatusUnauthorizedSignature';
                
                // Get the request data from the JSON file
                $requestData = Util::getRequest(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName
                );
                
                // Use the original partner reference from the successful transaction
                $requestData['originalPartnerReferenceNo'] = self::$originalPartnerReferencePaid;
                
                // Create headers with invalid signature to test authorization failure
                $headers = Util::getHeadersWithSignature(
                    'POST', 
                    '/rest/v1.0/emoney/topup-status',
                    $requestData,
                    true,
                    false,
                    true  // invalid_signature set to true
                );
                
                // Make direct API call with invalid signature
                try {
                    Util::executeApiRequest(
                        'POST',
                        'https://api.sandbox.dana.id/rest/v1.0/emoney/topup-status',
                        $headers,
                        $requestData
                    );
                    
                    $this->fail('Expected ApiException for invalid signature but the API call succeeded');
                } catch (ApiException $e) {
                    // We expect a 401 Unauthorized for invalid signature
                    $this->assertEquals(401, $e->getCode(), "Expected HTTP 401 Unauthorized for invalid signature, got {$e->getCode()}");

                    // Get the response body from the exception
                    $responseContent = (string)$e->getResponseBody();
                    
                    // Use assertFailResponse to validate the error response
                    Assertion::assertFailResponse(
                        self::$jsonPathFile, 
                        self::$titleCase, 
                        $caseName, 
                        $responseContent,
                        ['partnerReferenceNo' => self::$originalPartnerReferencePaid]
                    );
                } catch (Exception $e) {
                    $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
                }
            });
        } catch (Exception $e) {            
            $this->fail("Unexpected exception: " . $e->getMessage());
        }
    }
}